polkadot/erasure-coding/src/lib.rs

```diff
@@ -113,6 +138,45 @@ fn code_params(n_validators: usize) -> Result<CodeParams, Error> {
 	})
 }
 
+/// Reconstruct the v1 available data from the set of systematic chunks.
+///
+/// Provide a vector containing chunk data. If too few chunks are provided, recovery is not
+/// possible.
+pub fn reconstruct_from_systematic_v1(
+	n_validators: usize,
+	chunks: Vec<&[u8]>,
+) -> Result<AvailableData, Error> {
+	reconstruct_from_systematic(n_validators, chunks)
+}
+
+/// Reconstruct the available data from the set of systematic chunks.
+///
+/// Provide a vector containing the first k chunks in order. If too few chunks are provided,
+/// recovery is not possible.
+pub fn reconstruct_from_systematic<T: Decode>(
+	n_validators: usize,
+	chunks: Vec<&[u8]>,
+) -> Result<T, Error> {
+	let code_params = code_params(n_validators)?;
+	let k = code_params.k();
+
+	for chunk_data in chunks.iter().take(k) {
+		if chunk_data.len() % 2 != 0 {
+			return Err(Error::UnevenLength)
+		}
+	}
+
+	let bytes = code_params.make_encoder().reconstruct_from_systematic(
+		chunks
+			.into_iter()
+			.take(k)
+			.map(|data| WrappedShard::new(data.to_vec()))
```

Comment:
I've simplified the interface a bit in https://github.com/paritytech/erasure-coding/blob/main/src/lib.rs - take a look, this also gets rid of some unneeded clones.

Reply:
`WrappedShard` currently owns the data, and the reed-solomon lib mutates it, so it's not so easy to accept a slice here.

When working on https://github.com/paritytech/reed-solomon-novelpoly/issues/14, I got rid of `WrappedShard` completely, to minimise the differences between the rust and C++ impls. I didn't see any performance difference.

Personally I think it's out of scope for this PR (which is huge anyway) to optimise the interface, I wouldn't bother with this now - especially since early experiments didn't show an improvement.